File: samples/client/3_0_3_unit_test/java/src/main/java/org/openapijsonschematools/client/schemas/validation/UniqueItemsValidator.java
package org.openapijsonschematools.client.schemas.validation;

import org.openapijsonschematools.client.exceptions.ValidationException;
import org.checkerframework.checker.nullness.qual.Nullable;

import java.util.HashSet;
import java.util.List;
import java.util.Set;

public class UniqueItemsValidator implements KeywordValidator {
    public final boolean uniqueItems;

    public UniqueItemsValidator(boolean uniqueItems) {
        this.uniqueItems = uniqueItems;
    }

    @Override
    public @Nullable PathToSchemasMap validate(
        JsonSchema schema,
        @Nullable Object arg,
        ValidationMetadata validationMetadata,
        List<PathToSchemasMap> containsPathToSchemas
    ) {
        if (!(arg instanceof List)) {
            return null;
        }
        if (!uniqueItems) {
            return null;
        }
        Set<@Nullable Object> seenItems = new HashSet<>();
        for (@Nullable Object item: (List<?>) arg) {
            int startingSeenItemsSize = seenItems.size();
            seenItems.add(item);
            if (seenItems.size() == startingSeenItemsSize) {
                throw new ValidationException("Invalid list value, list contains duplicate items when uniqueItems is true");
            }
        }
        return null;
    }
}

File: samples/client/3_0_3_unit_test/kotlin/src/main/kotlin/org/openapijsonschematools/client/schemas/validation/UniqueItemsValidator.kt
package org.openapijsonschematools.client.schemas.validation

import org.openapijsonschematools.client.exceptions.ValidationException

class UniqueItemsValidator : KeywordValidator {
    @Throws(ValidationException::class)
    override fun validate(
        data: ValidationData
    ): PathToSchemasMap? {
        val uniqueItems: Boolean = data.schema.uniqueItems ?: return null
        if (data.arg !is List<*>) {
            return null
        }
        if (!uniqueItems) {
            return null
        }
        val seenItems: MutableSet<Any?> = HashSet()
        for (item in data.arg) {
            val startingSeenItemsSize = seenItems.size
            seenItems.add(item)
            if (seenItems.size == startingSeenItemsSize) {
                throw ValidationException("Invalid list value, list contains duplicate items when uniqueItems is true")
            }
        }
        return null
    }
}
